csrc/id_model/loop_promotion.cpp

```diff
@@ -116,47 +118,113 @@ bool isLoopGraphAlmostUniform(const IdModel& id_model) {
           id_model.idGraph(IdMappingMode::LOOP).toGroup(loop_id);
       const auto all_exact_groups =
           id_model.idGraph(IdMappingMode::EXACT).toGroups(*loop_group);
-      if (all_exact_groups.size() == 1) {
+      if (all_exact_groups.size() > 1) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+} // namespace
+
+ValGroups LoopPromotionMapBuilder::getInputGroupsOfExactGraph(
+    const ValGraph& exact_graph) const {
+  const std::unordered_set<TensorView*> included_tvs{
+      id_model_.tvs().begin(), id_model_.tvs().end()};
+
+  std::unordered_set<IterDomain*> non_input_ids;
+
+  for (auto tv_expr : id_model_.tvExprs()) {
+    for (const auto producer :
+         ir_utils::filterByType<TensorView>(tv_expr->inputs())) {
+      if (!included_tvs.count(producer)) {
```

Comment:
Is this always false?

Reply:
Hmm, that seems right. Not sure why I added this. Let me do some check

Reply:
Confirmed it's always false. Will remove it